File: android/src/main/java/app/capgo/capacitor/fastsql/CapgoCapacitorFastSqlPlugin.java
package app.capgo.capacitor.fastsql;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.util.HashMap;
import java.util.Map;

/**
 * Native SQL Plugin for Android
 *
 * This plugin uses a custom HTTP server for efficient data transfer,
 * bypassing Capacitor's standard bridge for better performance.
 */
@CapacitorPlugin(name = "CapgoCapacitorFastSql")
public class CapgoCapacitorFastSqlPlugin extends Plugin {

    private final String PLUGIN_VERSION = "7.2.2";

    private Map<String, SQLDatabase> databases = new HashMap<>();
    private SQLHTTPServer server;

    @PluginMethod
    public void connect(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        // Check if already connected
        if (databases.containsKey(database)) {
            if (server != null) {
                JSObject ret = new JSObject();
                ret.put("port", server.getPort());
                ret.put("token", server.getToken());
                ret.put("database", database);
                call.resolve(ret);
                return;
            }
        }

        try {
            // Get database path
            File dbFile = getDatabasePath(database);

            // Open database
            SQLDatabase db = new SQLDatabase(dbFile.getAbsolutePath());
            databases.put(database, db);

            // Start HTTP server if not already running
            if (server == null) {
                server = new SQLHTTPServer(databases);
                server.start();
            }

            JSObject ret = new JSObject();
            ret.put("port", server.getPort());
            ret.put("token", server.getToken());
            ret.put("database", database);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to connect to database: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        SQLDatabase db = databases.get(database);
        if (db == null) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        db.close();
        databases.remove(database);

        // Stop server if no more databases
        if (databases.isEmpty() && server != null) {
            server.stop();
            server = null;
        }

        call.resolve();
    }

    @PluginMethod
    public void getServerInfo(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        if (!databases.containsKey(database)) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        if (server == null) {
            call.reject("Server is not running");
            return;
        }

        JSObject ret = new JSObject();
        ret.put("port", server.getPort());
        ret.put("token", server.getToken());
        call.resolve(ret);
    }

    @PluginMethod
    public void execute(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        String statement = call.getString("statement");
        if (statement == null) {
            call.reject("Statement is required");
            return;
        }

        SQLDatabase db = databases.get(database);
        if (db == null) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        JSArray params = call.getArray("params", new JSArray());

        try {
            JSObject result = db.execute(statement, params);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to execute statement: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void beginTransaction(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        SQLDatabase db = databases.get(database);
        if (db == null) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        try {
            db.beginTransaction();
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to begin transaction: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void commitTransaction(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        SQLDatabase db = databases.get(database);
        if (db == null) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        try {
            db.commitTransaction();
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to commit transaction: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void rollbackTransaction(PluginCall call) {
        String database = call.getString("database");
        if (database == null) {
            call.reject("Database name is required");
            return;
        }

        SQLDatabase db = databases.get(database);
        if (db == null) {
            call.reject("Database '" + database + "' is not connected");
            return;
        }

        try {
            db.rollbackTransaction();
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to rollback transaction: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", PLUGIN_VERSION);
        call.resolve(ret);
    }

    private File getDatabasePath(String database) {
        File dataDir = getContext().getFilesDir();
        return new File(dataDir, database + ".db");
    }
}

File: ios/Sources/CapgoCapacitorFastSqlPlugin/CapgoCapacitorFastSqlPlugin.swift
import Foundation
import Capacitor
import SQLite3

/**
 * Native SQL Plugin for iOS
 *
 * This plugin uses a custom HTTP server for efficient data transfer,
 * bypassing Capacitor's standard bridge for better performance.
 */
@objc(CapgoCapacitorFastSqlPlugin)
public class CapgoCapacitorFastSqlPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.12"
    public let identifier = "CapgoCapacitorFastSqlPlugin"
    public let jsName = "CapgoCapacitorFastSql"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getServerInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "beginTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "commitTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rollbackTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private var databases: [String: SQLDatabase] = [:]
    private var server: SQLHTTPServer?

    @objc func connect(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        // Check if already connected
        if databases[database] != nil {
            if let server = server {
                call.resolve([
                    "port": server.port,
                    "token": server.token,
                    "database": database
                ])
                return
            }
        }

        do {
            // Get database path
            let dbPath = try getDatabasePath(database)

            // Open database
            let db = try SQLDatabase(path: dbPath)
            databases[database] = db

            // Start HTTP server if not already running
            if server == nil {
                server = try SQLHTTPServer(databases: databases)
                try server?.start()
            }

            guard let server = server else {
                call.reject("Failed to start HTTP server")
                return
            }

            call.resolve([
                "port": server.port,
                "token": server.token,
                "database": database
            ])
        } catch {
            call.reject("Failed to connect to database: \(error.localizedDescription)")
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard let db = databases[database] else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        db.close()
        databases.removeValue(forKey: database)

        // Stop server if no more databases
        if databases.isEmpty {
            server?.stop()
            server = nil
        }

        call.resolve()
    }

    @objc func getServerInfo(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard databases[database] != nil else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        guard let server = server else {
            call.reject("Server is not running")
            return
        }

        call.resolve([
            "port": server.port,
            "token": server.token
        ])
    }

    @objc func execute(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard let statement = call.getString("statement") else {
            call.reject("Statement is required")
            return
        }

        guard let db = databases[database] else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        let params = call.getArray("params", JSValue.self) ?? []

        do {
            let result = try db.execute(statement: statement, params: params)
            call.resolve(result)
        } catch {
            call.reject("Failed to execute statement: \(error.localizedDescription)")
        }
    }

    @objc func beginTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard let db = databases[database] else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        do {
            try db.beginTransaction()
            call.resolve()
        } catch {
            call.reject("Failed to begin transaction: \(error.localizedDescription)")
        }
    }

    @objc func commitTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard let db = databases[database] else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        do {
            try db.commitTransaction()
            call.resolve()
        } catch {
            call.reject("Failed to commit transaction: \(error.localizedDescription)")
        }
    }

    @objc func rollbackTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            call.reject("Database name is required")
            return
        }

        guard let db = databases[database] else {
            call.reject("Database '\(database)' is not connected")
            return
        }

        do {
            try db.rollbackTransaction()
            call.resolve()
        } catch {
            call.reject("Failed to rollback transaction: \(error.localizedDescription)")
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": pluginVersion
        ])
    }

    private func getDatabasePath(_ database: String) throws -> String {
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        guard let documentsDirectory = paths.first else {
            throw NSError(domain: "CapgoCapacitorFastSQL", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Could not find documents directory"
            ])
        }

        let dbPath = (documentsDirectory as NSString).appendingPathComponent("\(database).db")
        return dbPath
    }
}
